File: Python/543. Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:        
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        ans = 1
        def diameter(root):
            nonlocal ans
            if not root:
                return 0
            left = diameter(root.left)
            right = diameter(root.right)            
            ans = max(ans, left + right + 1)
            return max(left, right) + 1
        diameter(root)
        return ans - 1
        

File: 2021/543. Diameter of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int maxPath = 0;
        dfs(root, maxPath);
        return maxPath;        
    }
    
    int dfs(TreeNode* root, int& maxPath) {
        if (root == NULL) return 0;
        int left = dfs(root->left, maxPath);
        int right = dfs(root->right, maxPath);
        maxPath = max(left + right, maxPath);
        return max(left, right) + 1;
    }
};